### DISCUSSION TITLE
What is the most robust/reliable IR format for transmission?

### DISCUSSION THREAD

**Discussion by quicksketch** [Arduino-IRremote/Arduino-IRremote#1278] [Q&A] [answered]
Hello folks,

I'm creating two pieces of hardware. One is a transmitter and one a receiver. They are ESP32S3 based, 240Mhz with 8MB of SDRAM. The transmitter is a custom chip and the receiver is a [Seeed Studio XIAO ESP32S3](https://www.seeedstudio.com/XIAO-ESP32S3-p-5627.html). The Arduino-IRremote library will be used on both the sender and receiver, so I can choose the format to be used.

The transmitter operates via gyroscope/accelerometer, like a wand, so it's being waved around a bunch. The receiver is likely to be placed about 5 to 10 feet away. Because sending a command is going to take some time to input since it's done by the accelerometer rather than a single button push, I want to ensure the transmission is received because it's going to be frustrating to repeat if the command does not get received.

So I think what I'm looking for is a format that:

- Operates quickly with minimal transmission time.
- The receiver can be sure it's received a complete signal.
- The transmitter can repeat the signal for a span of about .5ms (or less) to give it time to ensure the receiver is targeted well enough.
- The receiver should only process the provided command a single time within a 1 second period. However, if a different command is sent, should be processed immediately.

NEC is obviously the most widely used protocol, but is it the most robust and reliable for such a situation? I'm curious what others would recommend in a situation where they get to choose the protocol.

**Comment by quicksketch** [comment, marked as answer]
Answering my own question from my research:

With this library, the [FAST Protocol](https://github.com/Arduino-IRremote/Arduino-IRremote?tab=readme-ov-file#the-fast-protocol) looks like it would be the best bet because:

- It has a very short data header, speeding up transfer time.
- It does not send a special "repeat code", instead the same command is repeated every time (same as "NEC2").
- In my situation, I don't need to assign a special address as there will only be one device present for this purpose (i.e. it's not in an entertainment system where there is a TV, Roku, DVD player, etc).

These factors would make FAST a well-suited protocol for my application.

Unfortunately, I need TWO receivers in my application, pointed in opposite directions that receive different signals. This library is [not capable of multiple receivers with separate signals](https://github.com/Arduino-IRremote/Arduino-IRremote?tab=readme-ov-file#multiple-ir-receiver-and-sender-instances) (only multiple receivers with a shared signal), so the whole question is moot; since I can't use this library at all.

Other than using multiple Arduinos (which I don't want to do for maintenance purposes), I'm switching to using a separate dedicated decoder chip which uses exclusively NEC. These are fairly rare, but these are available choices:

- https://electricdollarstore.com/remote.html (I2C)
- https://forum.hobbycomponents.com/viewtopic.php?t=3087 (I2C)
- https://www.amazon.com/Ganasome-Infrared-Transceiver-Decoder-Module/dp/B0CYZGTTBY (UART)

That said, I learned a lot using this library and although I won't be able to use it, I appreciate the vast resources that have gone into creating and maintaining it; including its expansive examples, support for different protocols, and wide device support. Thank you for the work on this project!

**Reply by ArminJo** [reply]
Have you ever build a system with 2 receivers in opposite directions?
You will most likely receive a signal at both receivers due to reflections. 
Try it with your remote and one receiver heading in the opposite direction.

**Reply by quicksketch** [reply]
I tried making multiple `IRrecv` instances, i.e.
```
IRrecv IrReceiver1;
IRrecv IrReceiver2;
```

But the data leaks between them. Both instances would report having the same signal even when one is listening on a disconnected PIN.

I **fully disconnected** one receiver from the circuit and the data still leaked into the other instance. So at least in this situation, it seems to be a software problem. I tried the [gerivega fork](https://github.com/gerivega/Arduino-IRremote) of this project which supposedly might help, but it also seemed to have data leakage at times. Not 100% of the time but still about 20% of the time.

**Reply by quicksketch** [reply]
>  You will most likely receive a signal at both receivers due to reflections.

Oh I realize that you're indicating my entire objective is not going to work. This is my first infrared system yes, but as this is for a special purpose in a controlled environment, reflections should be avoidable. If not, the perhaps addressing would be beneficial after all but I would prefer the same remote worked on both sides, but I can make do if I need to give them individual addresses.

**Reply by quicksketch** [reply]
The [Ganasome receiver/decoder](https://www.amazon.com/Ganasome-Infrared-Transceiver-Decoder-Module/dp/B0CYZGTTBY) is working perfectly for my purposes, FWIW. If I make multiple of these devices, it would be nice if it were possible to simplify the hardware. Now I've essentially got 3 microcontrollers (one for each IR receiver and one to run my actual program).

I think the general concept of needing multiple IR receivers is niche, yes, but it seems technically like a viable solution to the situation I have.
